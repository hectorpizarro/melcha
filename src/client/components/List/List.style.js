/**
 * Estilos para List.js
 */
import styled from "styled-components";

const Container = styled.div`
  min-width: 640px;
`;

// Imagen mostrada en cada fila
const Image = styled.img`
  width: 180px;
  height: 180px;
  border-radius: 4px;
`;

// Fila, contiene la seccion izquierda con foto y contenido, seccion media esta vacia para poder respetar el grid pero muestra border top y la seccion derecha muestra la ciudad.
const RowGrid = styled.div`
  color: ${(props) => props.theme.colors.black};
  display: grid;
  flex-grow: 1;
  grid-template-columns: 1fr 1rem 1fr 1rem 1fr 1rem 1fr 1rem 1fr 1rem 1fr 1rem 1fr 1rem 1fr 1rem 1fr 1rem 1fr 1rem 1fr 1rem 1fr;
  grid-template-areas: ". . areaContent areaContent areaContent areaContent areaContent areaContent areaContent areaContent areaContent areaContent areaContent areaContent areaGap areaGap areaGap areaGap areaLocation areaLocation areaLocation . .";

  & .section {
    background-color: ${(props) => props.theme.colors.white};
    display: flex;
    flex-direction: column;
  }

  & .contentSection {
    grid-area: areaContent;
    padding: 16px 0px 0px 16px;

    &.first {
      padding: 0px 0px 0px 16px;
      border-top-left-radius: 4px;
    }
    &.last {
      padding: 16px 0px 16px 16px;
    }
  }

  & .middleSection {
    grid-area: areaGap;
    padding-top: 16px;

    &.first {
      padding-top: 0px;
    }
    &.last {
      padding-top: 16px;
    }
  }

  & .locationSection {
    grid-area: areaLocation;
    padding: 16px 16px 0px 0px;

    &.first {
      padding: 0px 16px 0px 0px;
      border-top-right-radius: 4px;
    }
    &.last {
      padding: 16px 16px 16px 0px;
    }
  }
`;

// contenido de la seccion izquierda
const LeftColumn = styled.div`
  border-top: ${(props) =>
    props.showTopBorder ? "0" : `1px solid ${props.theme.colors.graylight}`};
  display: flex;
  flex: 1;
  padding-top: 16px;
`;

// contenido de la seccion central
const MiddleColumn = styled.div`
  border-top: ${(props) =>
    props.showTopBorder ? "0" : `1px solid ${props.theme.colors.graylight}`};
`;

// boton usado como wrap de la foto y titulo
const LinkButton = styled.button`
  background-color: transparent;
  border: 0;
  cursor: pointer;
  display: flex;
  font-size: 18px;
  margin: 0;
  padding: 0;
  text-align: left;
  outline: none;
`;

// contenido de la seccion derecha
const RightColumn = styled.div`
  border-top: ${(props) =>
    props.showTopBorder ? "0" : `1px solid ${props.theme.colors.graylight}`};
  flex: 1;
  font-size: 12px;
  padding-top: 32px;
`;

// Grid para los textos a la dewrecha de la foto: currency, amount, icono para free shipping y title.
const ContentGrid = styled.div`
  display: grid;
  padding: 16px;
  grid-template-columns: auto auto 1fr;
  grid-template-rows: auto 1fr;
  grid-template-areas:
    "areaCurrency areaAmount areaFreeShip"
    "areaTitle areaTitle areaTitle";
  font-size: 18px;
  grid-column-gap: 8px;
  grid-row-gap: 16px;

  .currency {
    grid-area: areaCurrency;
  }
  .amount {
    grid-area: areaAmount;
  }
  .free_shipping {
    padding: 0;
    grid-area: areaFreeShip;
    display: flex;
  }
  .title {
    grid-area: areaTitle;
  }
`;

export default {
  Container,
  Image,
  RowGrid,
  LeftColumn,
  MiddleColumn,
  LinkButton,
  RightColumn,
  ContentGrid,
};
