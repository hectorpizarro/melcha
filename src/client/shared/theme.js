/**
 * Theme usado en todos los estilos, inyectado por el ThemeProvider de styled components.
 */

export default {
  colors: {
    white: "#ffffff",
    yellow: "#ffe600",
    black: "#333333",
    graydark: "#666666",
    gray: "#999999",
    graylight: "#eeeeee",
    blue: "#3483FA",
  },
};
