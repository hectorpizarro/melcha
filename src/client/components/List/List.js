/**
 * Muestra listado de items cuando usuario hace busqueda.
 */
import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { useHistory } from "react-router-dom";

import { getBreadcrumb, getNumLocale } from "../../shared/utils";
import Styled from "./List.style";
import ImgFreeShipping from "./images/ic_shipping@2x.png.png.png";
import { setLoading, selectItem } from "../../redux/itemReducer";
import Loader from "../Loader/Loader";

export default () => {
  const history = useHistory();
  const dispatch = useDispatch();
  // Carga list para el render, loading es flag para mostrar loader.
  const {
    root: { list, loading },
  } = useSelector((state) => state);

  if (loading) {
    // flag indica que listado aun no esta procesado, mostrar loader
    return <Loader />;
  }

  /**
   * Activa el flag para mostrar loader, luego ejectua async selectItem para cargar el item, finalmente redirecciona la app al route de item detail.
   * @param {String} id - Item id
   */
  const gotoDetail = async (id) => {
    await dispatch(setLoading({ loading: true }));
    dispatch(selectItem(id));
    history.push(`/items/${id}`);
  };

  // total de items en la lista, usado en la logica para manejar estilos
  const len = list.data.items.length;

  // Array de categorias a mostrar
  const topCats = getBreadcrumb(list.data.categories);

  return (
    <div>
      <Styled.Breadcrumb>
        <div className="breadCrumb">
          {/* Si el listado esta vacio mostrar mensaje */}
          {list.data.items.length === 0 && (
            <span>No hay resultados en su b&uacute;squeda.</span>
          )}
          {/* Muestra breadcrumb de categorias */}
          {topCats.map((cat, idx) => {
            return (
              <React.Fragment key={cat}>
                {idx > 0 && " > "}
                {idx === topCats.length - 1 ? (
                  <strong>{cat}</strong>
                ) : (
                  <span>{cat}</span>
                )}
              </React.Fragment>
            );
          })}
        </div>
      </Styled.Breadcrumb>
      {list.data.items.map((item, idx) => (
        <div key={item.id}>
          <Styled.RowGrid>
            <div
              className={`section contentSection ${idx === 0 ? "first" : ""} ${
                idx === len - 1 ? "last" : ""
              }`}
            >
              <Styled.LeftColumn showTopBorder={idx === 0}>
                <Styled.LinkButton onClick={() => gotoDetail(item.id)}>
                  <Styled.Image src={item.picture} alt="" />
                </Styled.LinkButton>
                <Styled.ContentGrid>
                  <div className="currency">{item.price.currency}</div>
                  <div className="amount">
                    {getNumLocale(item.price.amount)}
                  </div>
                  <div className="free_shipping">
                    {item.free_shipping && (
                      <Styled.IconFreeShip src={ImgFreeShipping} alt="" />
                    )}
                  </div>
                  <div className="title">
                    <Styled.LinkButton onClick={() => gotoDetail(item.id)}>
                      <span>{item.title}</span>
                    </Styled.LinkButton>
                  </div>
                </Styled.ContentGrid>
              </Styled.LeftColumn>
            </div>
            <div
              className={`section locationSection ${idx === 0 ? "first" : ""} ${
                idx === len - 1 ? "last" : ""
              }`}
            >
              <Styled.RightColumn showTopBorder={idx === 0}>
                {item.city}
              </Styled.RightColumn>
            </div>
            <div
              className={`section middleSection ${idx === 0 ? "first" : ""} ${
                idx === len - 1 ? "last" : ""
              }`}
            >
              <Styled.MiddleColumn showTopBorder={idx === 0} />
            </div>
          </Styled.RowGrid>
        </div>
      ))}
    </div>
  );
};
