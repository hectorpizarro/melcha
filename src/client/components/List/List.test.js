import React from "react";
import { Provider } from "react-redux";
import { ThemeProvider } from "styled-components";
import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";
import { mount } from "enzyme";
import renderer from "react-test-renderer";

import theme from "../../shared/theme";
import List from "./List";

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
let store = null; // es global para usar dentro de CompWithProviders

const CompWithProviders = () => (
  <Provider store={store}>
    <ThemeProvider theme={theme}>
      <List />
    </ThemeProvider>
  </Provider>
);

describe("<List />", () => {
  it("Renders without crashing", () => {
    store = mockStore({
      root: {
        firstLoaded: false,
        searchTerm: "",
        list: null,
        loading: true,
      },
    });
    const wrapper = mount(<CompWithProviders />);
    expect(wrapper.text()).toEqual("");
  });
  it("Check snapshot", () => {
    store = mockStore({
      root: {
        firstLoaded: false,
        searchTerm: "",
        list: null,
        loading: true,
      },
    });
    const tree = renderer.create(<CompWithProviders />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("Renders with no list data", () => {
    store = mockStore({
      root: {
        firstLoaded: false,
        searchTerm: "",
        list: null,
        loading: false,
      },
    });
    const wrapper = mount(<CompWithProviders />);
    expect(wrapper.text()).toEqual("No hay resultados en su bÃºsqueda.");
  });
  it("Check snapshot with no list data", () => {
    store = mockStore({
      root: {
        firstLoaded: false,
        searchTerm: "",
        list: null,
        loading: false,
      },
    });
    const tree = renderer.create(<CompWithProviders />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Renders with list data", () => {
    store = mockStore({
      root: {
        firstLoaded: false,
        searchTerm: "",
        list: {
          data: {
            items: [
              {
                id: 1,
                picture: "src",
                price: { currency: "ARS", amount: 1122 },
                free_shipping: true,
                title: "mock-title",
                city: "BA",
              },
            ],
            categories: ["aa", "bb", "cc"],
          },
        },
        loading: false,
      },
    });
    const wrapper = mount(<CompWithProviders />);
    expect(wrapper.text()).toEqual("cc > bb > aaARS1.122mock-titleBA");
  });

  it("Check snapshot with list data", () => {
    store = mockStore({
      root: {
        firstLoaded: false,
        searchTerm: "",
        list: {
          data: {
            items: [
              {
                id: 1,
                picture: "src",
                price: { currency: "ARS", amount: 1122 },
                free_shipping: true,
                title: "mock-title",
                city: "BA",
              },
            ],
            categories: ["aa", "bb", "cc"],
          },
        },
        loading: false,
      },
    });
    const tree = renderer.create(<CompWithProviders />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
