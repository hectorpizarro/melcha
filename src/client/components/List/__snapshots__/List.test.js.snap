// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> Check snapshot 1`] = `
<div
  className="sc-fubCfw epvmuM"
  height="inherit"
>
  <div
    className="lds-roller"
  >
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
  </div>
</div>
`;

exports[`<List /> Check snapshot with list data 1`] = `
<div
  className="sc-bdfBwQ iMePgP"
>
  <div
    className="sc-pFZIQ kMoonX"
  >
    <div
      className="breadCrumb"
    >
      <div>
        <span>
          cc
        </span>
         &gt; 
        <span>
          bb
        </span>
         &gt; 
        <strong>
          aa
        </strong>
      </div>
    </div>
  </div>
  <div>
    <div
      className="sc-dlfnbm dmdQNq"
    >
      <div
        className="section contentSection first last"
      >
        <div
          className="sc-hKgILt fQVHkt"
        >
          <button
            aria-label="Ver item"
            className="sc-jSgupP caFwzH"
            onClick={[Function]}
          >
            <img
              alt=""
              className="sc-gsTCUz dDhmPB"
              src="src"
            />
          </button>
          <div
            className="sc-iBPRYJ cQqOqG"
          >
            <div
              className="currency"
            >
              ARS
            </div>
            <div
              className="amount"
            >
              1.122
            </div>
            <div
              className="free_shipping"
            >
              <img
                alt=""
                className="sc-jrAGrp eUEZkx"
                src="ic_shipping@2x.png.png.png"
              />
            </div>
            <div
              className="title"
            >
              <button
                aria-label="Ver item"
                className="sc-jSgupP caFwzH"
                onClick={[Function]}
              >
                <span>
                  mock-title
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="section locationSection first last"
      >
        <div
          className="sc-gKsewC cQWlqB"
        >
          BA
        </div>
      </div>
      <div
        className="section middleSection first last"
      >
        <div
          className="sc-eCssSg itivUS"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<List /> Check snapshot with no list data 1`] = `
<div
  className="sc-bdfBwQ iMePgP"
>
  <div
    className="sc-pFZIQ kMoonX"
  >
    <div
      className="breadCrumb"
    >
      <div>
        <span>
          No hay resultados en su b√∫squeda.
        </span>
      </div>
    </div>
  </div>
</div>
`;
